tests/model_fields/test_filefield.py

```diff
@@ -147,3 +147,11 @@ def test_pickle(self):
                         self.assertEqual(document.myfile.field, loaded_myfile.field)
                     finally:
                         document.myfile.delete()
+
+    def test_model_attribute(self):
+        """The model attribute of file fields should reference the concrete
+        model even if the field was defined on an abstract model
+
+        Regression test for #31812"""
+        field = Document._meta.get_field("myfile")
+        self.assertEqual(field.model, Document)  # Not AbstractDocument
```

Comment:
This works even without a patch so we should switch to:
```python
document = MyDocument(myfile='test_file.py')
self.assertEqual(document.myfile.field.model, MyDocument)
```
We could also move new models to the test, e.g.
```python
    @isolate_apps('model_fields')
    def test_abstract_filefield_model(self):
        class AbstractMyDocument(models.Model):
            myfile = models.FileField(upload_to='unused', unique=True)

            class Meta:
                abstract = True

        class MyDocument(AbstractMyDocument):
            pass
    
        document = MyDocument(myfile='test_file.py')
        self.assertEqual(document.myfile.field.model, MyDocument)
```